src/canvasLogic/Canvas.ts

```diff
@@ -0,0 +1,175 @@
+            default:
+                throw new IncorrectShapeError(this.selectedShape);
+            }
+            break;
+        case Tools.Text:
+            this.currentModifiedObject = new Text(event.x, event.y);
+            break;
+        default:
+            this.currentModifiedObject = null;
+            if (
+                this.selectedTool !== Tools.Cursor
+          && this.selectedTool !== Tools.Pen
+            ) {
+                throw new IncorrectToolError(this.selectedTool);
+            }
+        }
+    }
+
+    public move(event: IMouseEvent): void {
+        if (!this.currentModifiedObject) return;
+
+        switch (this.selectedTool) {
+        case Tools.Shapes:
+            if (!(this.currentModifiedObject instanceof Shape)) return;
+            this.currentModifiedObject.change(event.x, event.y);
+            break;
+        default:
+            this.currentModifiedObject = null;
+            if (
+                this.selectedTool !== Tools.Cursor
+          && this.selectedTool !== Tools.Text
+          && this.selectedTool !== Tools.Pen
+            ) {
+                throw new IncorrectToolError(this.selectedTool);
+            }
+        }
+    }
+
+    public up(event: IMouseEvent): void {
+        if (!this.currentModifiedObject) return;
+
+        switch (this.selectedTool) {
+        case Tools.Shapes:
+            if (!(this.currentModifiedObject instanceof Shape)) return;
+            this.currentModifiedObject.end(event.x, event.y);
+            break;
+        default:
+            this.currentModifiedObject = null;
+            if (
+                this.selectedTool !== Tools.Cursor
+          && this.selectedTool !== Tools.Text
+          && this.selectedTool !== Tools.Pen
+            ) {
+                throw new IncorrectToolError(this.selectedTool);
+            }
+        }
+    // TODO: Make it work (send an event to the parent)
+    // if ( this.selectedTool !== Tools.Pen ) this.changeTool(Tools.Cursor);
+    }
+
+    public get currentObject(): CanvasObject | null {
+        return this.currentModifiedObject;
+    }
+
+    public get currentTool(): Tools_ {
+        return this.selectedTool;
+    }
+}
+
+export abstract class Canvas {
+    public id: CanvasId;
+
+    protected _canvasMouse: CanvasMouse;
+
+    protected _conference: Conference | undefined;
+
+    constructor(id: CanvasId, conference?: Conference) {
+        this.id = id;
+
+        this._canvasMouse = new CanvasMouse();
+        this._conference = conference;
+    }
+
+    abstract update(data: string): void
+
+    public set canvasMouse(canvasMouse: CanvasMouse) {
+        this._canvasMouse = canvasMouse;
+    }
+
+    public set conference(conference: Conference) {
+        this._conference = conference;
+    }
+
+    public sendToConference(data: object): void {
+        console.log("Sending to conference", {"target": this.id, ...data});
+        this._conference?.send({"target": this.id, ...data});
```

Comment:
It would be awesome if there were log levels of some sort (DEBUG, INFO, WARNING, ERROR, etc.), but it's not that crucial rn

Reply:
I think so, but in my opinion debug and other verbose levels must be sent to be logged on backend and it is a very big feature, which needs independent branch